koku/masu/processor/gcp/gcp_report_processor.py

```diff
@@ -60,17 +95,120 @@ def __init__(self, schema_name, report_path, compression, provider_uuid, manifes
+        self._manifest_id = manifest_id
+        self._provider_uuid = provider_uuid
 
         self._schema = schema_name
 
+        with GCPReportDBAccessor(self._schema) as report_db:
+            self.report_schema = report_db.report_schema
+            self.existing_bill_map = report_db.get_cost_entry_bills()
+            self.existing_product_map = report_db.get_products()
+            self.existing_projects_map = report_db.get_projects()
+
         LOG.info("Initialized report processor for file: %s and schema: %s", report_path, self._schema)
 
         self.line_item_columns = None
 
+    def _get_line_item_type(self, row):
+        """Given a row find the line item type."""
+        item_type = "other"
+        service_alias = row.get("service.description", "")
+        key_list = list(GCP_SERVICE_LINE_ITEM_TYPE_MAP.keys())
+        if service_alias in key_list:
+            return GCP_SERVICE_LINE_ITEM_TYPE_MAP[service_alias]
+        service_alias = service_alias.replace(" ", "").lower()
+        for key in key_list:
+            if service_alias == key.replace(" ", "").lower():
+                item_type = GCP_SERVICE_LINE_ITEM_TYPE_MAP[key]
+                break
+        return item_type
+
+    def _delete_line_items_in_range(self, bill_id, scan_start):
+        """Delete stale data between date range."""
+        gcp_date_filter = {"usage_start__gte": scan_start}
+
+        if not self._manifest_id:
+            return False
+        with ReportManifestDBAccessor() as manifest_accessor:
+            num_processed_files = manifest_accessor.number_of_files_processed(self._manifest_id)
+            if num_processed_files != 0:
+                return False
+
+        with GCPReportDBAccessor(self._schema) as accessor:
+            line_item_query = accessor.get_lineitem_query_for_billid(bill_id)
+            line_item_query = line_item_query.filter(**gcp_date_filter)
+            log_statement = (
+                f"Deleting data for:\n"
+                f" schema_name: {self._schema}\n"
+                f" provider_uuid: {self._provider_uuid}\n"
+                f" bill ID: {bill_id}\n"
+                f" on or after {scan_start}."
+            )
+            LOG.info(log_statement)
+            line_item_query.delete()
+
+        return True
+
+    def _process_tags(self, row):
+        """Return a JSON string of AWS resource tags.
+
+        Args:
+            row (dict): A dictionary representation of a CSV file row
+
+        Returns:
+            (str): A JSON string of AWS resource tags
+
+        """
+        tags_dict = {}
+        labels = row.get("labels", [])
+        if isinstance(labels, str):
+            labels = json.loads(labels.replace("'", '"'))
+        for label in labels:
+            key = label["key"]
+            value = label["value"]
+            tags_dict[key] = value
+        return json.dumps(tags_dict)
+
+    def _get_usage_type(self, row):
+        """Return instance_type if there is one.
```

Comment:
More comment updates :)

Reply:
@dccurtis Starting to see how bad I am at copying and pasting 😂 

I will push up these changes soon, I am going to do it as one commit since I am going to remove the finalize report method. That was a good catch.